File: backend/auth-service/src/services/display_name.ts
export const DISPLAY_NAME_MIN_LENGTH = 2;
export const DISPLAY_NAME_MAX_LENGTH = 50;

export class DisplayNameValidationError extends Error {
  readonly code = 'VALIDATION_ERROR';

  constructor(message: string) {
    super(message);
    this.name = 'DisplayNameValidationError';
  }
}

const CONTROL_CHARS = /[\u0000-\u001F\u007F-\u009F]/;
const UNPAIRED_SURROGATE = /[\uD800-\uDFFF]/;
const LETTERS = /\p{L}/gu;

/**
 * Normalize and validate a passenger display name.
 *
 * Server-authoritative. Rejects empty/whitespace, control characters,
 * unpaired surrogates, and lengths outside [2, 50] after NFC + trim.
 */
export function normalizeDisplayName(raw: unknown): string {
  if (typeof raw !== 'string') {
    throw new DisplayNameValidationError('displayName must be a string.');
  }

  const normalized = raw.normalize('NFC').trim().replace(/\s+/g, ' ');

  if (normalized.length === 0) {
    throw new DisplayNameValidationError('displayName is required.');
  }
  if (CONTROL_CHARS.test(normalized) || UNPAIRED_SURROGATE.test(normalized)) {
    throw new DisplayNameValidationError(
      'displayName contains invalid characters.',
    );
  }
  if (normalized.length < DISPLAY_NAME_MIN_LENGTH) {
    throw new DisplayNameValidationError(
      `displayName must be at least ${DISPLAY_NAME_MIN_LENGTH} characters.`,
    );
  }
  if (normalized.length > DISPLAY_NAME_MAX_LENGTH) {
    throw new DisplayNameValidationError(
      `displayName must be at most ${DISPLAY_NAME_MAX_LENGTH} characters.`,
    );
  }

  const letterCount = normalized.match(LETTERS)?.length ?? 0;
  if (letterCount < DISPLAY_NAME_MIN_LENGTH) {
    throw new DisplayNameValidationError(
      'displayName must contain at least two letters.',
    );
  }

  return normalized;
}

File: mobile/lib/core/utils/display_name.dart
/// Passenger display-name rules (must match auth-service `display_name.ts`).
///
/// Server remains authoritative; this is fail-fast UX only.
class DisplayName {
  DisplayName._();

  static const minLength = 2;
  static const maxLength = 50;

  static final _controlChars = RegExp(r'[\u0000-\u001F\u007F-\u009F]');
  static final _letters = RegExp(r'\p{L}', unicode: true);

  /// Returns a normalized name, or a user-facing error string.
  static String? validationError(String raw) {
    try {
      normalize(raw);
      return null;
    } on FormatException catch (e) {
      return e.message;
    }
  }

  static String normalize(String raw) {
    final normalized = raw.trim().replaceAll(RegExp(r'\s+'), ' ');
    if (normalized.isEmpty) {
      throw const FormatException('Please enter your name.');
    }
    if (_controlChars.hasMatch(normalized)) {
      throw const FormatException('Name contains invalid characters.');
    }
    if (normalized.length < minLength) {
      throw const FormatException('Name must be at least 2 characters.');
    }
    if (normalized.length > maxLength) {
      throw const FormatException('Name must be at most 50 characters.');
    }
    final letterCount = _letters.allMatches(normalized).length;
    if (letterCount < minLength) {
      throw const FormatException('Name must contain at least two letters.');
    }
    return normalized;
  }
}
